### DISCUSSION TITLE
Issue with cy.origin() and otplib Integration in Cypress Tests

### DISCUSSION THREAD

**Discussion by Padmini-bulnda** [cypress-io/cypress#30646] [Questions and Help]
Hello Cypress Community,

I’m facing an issue while trying to use cy.origin() in combination with otplib for 2-factor authentication in my Cypress tests. Despite enabling the experimentalOriginDependencies flag, I'm encountering errors related to process is not defined and challenges with passing dependencies correctly.

Here's the relevant part of my code:
import data from "../../fixtures/user.json";

class LoginPage {
    webLocators = {
        username: '#awsui-input-0',
        password: '#awsui-input-1',
        nextbtn: '.awsui-button',
        submitbtn: 'button[type="submit"]',
        Authenticatorbtn: "input[value='TOTP']",
        Pointer: "a[class='pointer']",
        secretkeylink: "span[class='secret-key awsui-util-mr-s']",
        Inputsecter: 'input[type="text"]'
    }

    loginWithUI() {
        cy.wait(7000);
        try {
            cy.origin('https://ap-south-1.signin.aws', { 
                args: { webLocators: this.webLocators, data: { username: data.username, password: data.password } },
            }, (args) => {
                const otplib = Cypress.require('otplib');
                const _ = Cypress.require('lodash');

                cy.visit('https://propeli-qa.sfxapps.com/dashboard');
                cy.get(args.webLocators.username, { timeout: 12000 }).should('be.visible');
                cy.get(args.webLocators.username).type(args.data.username);
                cy.get(args.webLocators.nextbtn).click();
                cy.get(args.webLocators.password).type(args.data.password);
                cy.get(args.webLocators.submitbtn).click();
                cy.wait(20000);
                cy.get(args.webLocators.Authenticatorbtn).click();
                cy.get(args.webLocators.submitbtn).click();
                cy.get(args.webLocators.Pointer).click();
                cy.get(args.webLocators.secretkeylink).invoke('text').then((secretKey) => {
                    console.log("this is " + secretKey);
                    const secret = otplib.authenticator.generate(secretKey);
                    cy.get(args.webLocators.Inputsecter).type(secret);
                    cy.get(args.webLocators.submitbtn).click();
                    cy.get(args.webLocators.submitbtn).click();
                    cy.log("success");
                });
            });

            cy.wait(20000);  // Adjust wait time and additional operations here as needed

        } catch (error) {
            console.error('Login failed:', error); // Log the error for debugging
            throw error;
        }
    }
}
export default new LoginPage();

I've made sure that:

- The experimentalOriginDependencies flag is enabled in cypress.json.
- Cypress.require() is used inside the cy.origin() callback.
- Despite these efforts, the process is not defined error persists. 
 
Any insights or suggestions on how to properly pass dependencies and handle this issue within the cy.origin() block would be greatly appreciated.

Thanks in advance for your help!

**Comment by portyu9** [comment]
Hi @Padmini-bulnda 

`Cypress.require()` inside `cy.origin()` is browser-targeted. The `process is not defined` error strongly suggests the version/build of `otplib` being bundled there expects Node globals.

I would move TOTP generation to the Node side with a `cy.task()` instead of loading `otplib` inside `cy.origin()`.

```ts
// cypress.config.ts
import { defineConfig } from 'cypress'
import { generate } from 'otplib'

export default defineConfig({
  e2e: {
    setupNodeEvents(on) {
      on('task', {
        async generateTotp(secret: string) {
          return generate({ secret })
        },
      })
    },
  },
})
```

Yield the secret from the origin block, generate the token in Node, then pass the resulting string back into another `cy.origin()` call:

```js
cy.origin(authOrigin, { args: { selectors } }, ({ selectors }) => {
  return cy.get(selectors.secretKey)
    .invoke('text')
    .then((text) => text.trim())
})
.then((secret) => cy.task('generateTotp', secret))
.then((otp) => {
  cy.origin(authOrigin, { args: { selectors, otp } }, ({ selectors, otp }) => {
    cy.get(selectors.otp).type(otp)
  })
})
```

If you are on `otplib` v12, the Node-side API is different (`authenticator.generate(secret)`); the example above is for current `otplib` v13.

One other issue in the sample: the callback is declared for the AWS origin but then calls `cy.visit()` for a different origin. On current Cypress, interactions on that other origin need their own matching `cy.origin()` context.

Please mark this answer as accepted if it helped, thank you.
